osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -1,181 +1,294 @@
         List<TypedSecret> typedSecretList = typedSecrets.getTypedSecret();
-
-        return typedSecretList.stream().filter(typedSecret -> typedSecret.getType() == secretType).findFirst();
+        this.validateGenerateAndStoreResponse(keyTypes, response, typedSecretList);
+        return this.convertSoapSecretsToSecretMapByType(typedSecrets.getTypedSecret());
     }
 
-    private GetSecretsRequest getSoapRequestForKey(String deviceIdentification, SecretType secretType) {
-        GetSecretsRequest request = new GetSecretsRequest();
+    private GenerateAndStoreSecretsRequest createGenerateAndStoreSecretsRequest(String deviceIdentification,
+            SecretTypes secretTypes) {
+        GenerateAndStoreSecretsRequest request = new GenerateAndStoreSecretsRequest();
         request.setDeviceId(deviceIdentification);
-        request.setSecretTypes(new SecretTypes());
-        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
-        secretTypeList.add(secretType);
+        request.setSecretTypes(secretTypes);
         return request;
     }
 
-    private byte[] decryptSoapSecret(String deviceIdentification, TypedSecret typedSecret) {
+    private void validateGenerateAndStoreResponse(List<SecurityKeyType> keyTypes,
+            GenerateAndStoreSecretsResponse response, List<TypedSecret> typedSecretList) {
+        this.validateOsgpResultAndTypedSecrets(response.getResult(), response.getTechnicalFault(),
+                response.getTypedSecrets(), keyTypes.size());
+        typedSecretList.forEach(ts -> {
+            if (ts.getSecret() == null) {
+                throw new IllegalStateException(String.format("Generated a NULL key (key_type=%s)", ts.getType()));
+            }
+        });
+    }
+
+    private void validateOsgpResultAndTypedSecrets(OsgpResultType result, Object fault, TypedSecrets typedSecrets,
+            int expectedNrKeys) {
+        if (!OsgpResultType.OK.equals(result)) {
+            throw new IllegalStateException("Could not generate and store keys in secret-mgmt: " + fault.toString());
+        } else if (typedSecrets == null || typedSecrets.getTypedSecret() == null) {
+            throw new IllegalStateException("No secrets in response");
+        } else if (expectedNrKeys != typedSecrets.getTypedSecret().size()) {
+            throw new IllegalStateException(
+                    String.format("Unexpected number of secrets in response: expected %s but " + "found %s",
+                            expectedNrKeys, typedSecrets.getTypedSecret().size()));
+        }
+    }
+
+    private byte[] decryptSoapSecret(TypedSecret typedSecret, boolean exceptionOnNull) {
+        boolean nullValue = typedSecret.getSecret() == null || typedSecret.getSecret().isEmpty();
+        if (exceptionOnNull && nullValue) {
+            throw new IllegalArgumentException("Cannot decrypt NULL value");
+        } else if (!exceptionOnNull && nullValue) {
+            return null;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1168)